Add initial ZeroHashMap

pdogr opened the pull request:
#2532 
Adds core functionality of ZeroHashMap. 
Perfect hashes are computed using [CHD](http://cmph.sourceforge.net/papers/esa09.pdf) algorithm, with [aHash](https://github.com/tkaitchuck/aHash) being used as the hashing function.

sffc commented on utils/zerovec/Cargo.toml:
```diff
@@ -31,6 +31,7 @@ yoke = { version = "0.6.0", path = "../yoke", optional = true }
 zerofrom = { version = "0.1.0", path = "../zerofrom" }
 zerovec-derive = {version = "0.8.0", path = "./derive", optional = true}
 databake = { version = "0.1.0", path = "../../utils/databake", features = ["derive"], optional = true }
+ahash = { version = "0.7.6" }
```
Three comments on `ahash`:

1. It should be an optional dependency. A feature such as "hashmap" should enable it.
2. Please include it with `default_features = false` in order to disable the "std" feature.
3. The docs for `ahash` say: "AHash is a hashing algorithm is intended to be a high performance, (hardware specific), keyed hash function." Can you verify whether AHash produces stable results on all architectures? If it is "hardware specific" then it might not be suitable for use in a serialization library. One of the main value propositions of zerovec is that we work the same on all processors.

robertbastian replied:
> Because it does not have a fixed standard, different computers or computers on different versions of the code will observe different hash values. As such, aHash is not recommended for use other than in-memory maps. Specifically, aHash is not intended for network use or in applications which persist hashed values. (In these cases HighwayHash would be a better choice)

- https://github.com/tkaitchuck/ahash#goals-and-non-goals

pdogr replied:
There are platform independent hashes as well like [seahash](https://docs.rs/seahash/latest/seahash/), [xxHash](https://github.com/Cyan4973/xxHash).

robertbastian replied:
or [HighwayHash](https://docs.rs/highway/0.8.0/highway/) as mentioned

sffc replied:
Two non-negotiables:

1. Should be `no_std`
2. Should be stable across versions and processors

And other priorities to consider:

1. Small code size
2. Robust against DOS collisions
3. High performance

pdogr replied:
Using `t1ha1 hash as discussed.

sffc commented on utils/zerovec/src/map/map.rs:
```diff
@@ -292,6 +292,11 @@ where
             &'b <K as ZeroMapKV<'a>>::GetType,
             &'b <V as ZeroMapKV<'a>>::GetType,
         ),
+    > + ExactSizeIterator<
```
Nit: These changes in `map` could stand by themselves. Consider putting them in their own PR.

pdogr replied:
Created PR #2580 for this.

robertbastian replied:
Please sync

sffc commented on utils/zerovec/src/map/hashmap.rs:
```diff
@@ -0,0 +1,270 @@
+
+#[derive(Debug, Clone)]
+pub struct HashIndex<'a> {
+    displacements: ZeroVec<'a, u32>,
+    reverse_mapping: ZeroVec<'a, u32>,
```
Suggestion: use `FlexZeroVec` so that you can store and read `usize` values (and also save some space). It has a small lookup penalty, but this should be minor relative to the hash calculation.

sffc commented on utils/zerovec/src/map/hashmap.rs:
```diff
@@ -0,0 +1,270 @@
+    V: ZeroMapKV<'a> + ?Sized,
+{
+    index: HashIndex<'a>,
+    keys: K::Container,
```
Suggestion: We don't need the keys for lookup, so we shouldn't store them. We won't be able to have `iter_keys` but that's fine. If necessary we could have a `ZeroHashMapWithKeys` if someone needs this (or just encourage them to use `ZeroMap` if they care about iterating over the keys).

pdogr replied:
Removed keys from `ZeroHashMapStatic`.

robertbastian replied:
Huh, this is not how hashmaps usually work. What about hash collisions? We should have to compare to the actual key somewhere, unless this algorithm somehow encodes unlimited entropy in two `u32`s.

pdogr replied:
Should we have two variants then as if we support unknown keys it would lead to a comparison of the key on every lookup.

robertbastian replied:
Long term yes, but for now we should focus on the general case, the key-less one can be an optimization for later.

sffc replied:
It's a discussion to be had. Given the multi-stage lookup, it seems that hash collisions are unlikely. It would be reasonable to argue that it's okay to (rarely) get GIGO results, so long as that possibility is documented, and if you really care about avoiding collisions, use ZeroMap.

robertbastian replied:
Perfect minimal hash functions mean that *every* key that is not in the map collides with some key in the map. It's not that some keys might collide with probability 2^-32 or something, the codomain of the hash function is the size of the collection so without storing the keys we will not know which keys are in the map and which aren't.

pdogr replied:
Yes if the first hash matches you are guaranteed a collision

sffc replied:
OK. You're right. I should read the paper.

robertbastian replied:
Take a look at http://stevehanov.ca/blog/?id=119

sffc reviewed (commented):
Praise: This looks like the type of hash map impl that will work well in ICU4X. It looks algorithmically similar to the approach rkyv is taking.

I would like to see benches against ZeroMap before merging, both data size and lookup performance, and code size for bonus points.

pdogr commented:
> Praise: This looks like the type of hash map impl that will work well in ICU4X. It looks algorithmically similar to the approach rkyv is taking.
> 
> I would like to see benches against ZeroMap before merging, both data size and lookup performance, and code size for bonus points.

Yes I based my implementation on rkyv's hashmap with a few changes. Both use CHD algorithm with the random hash function approach. There is a practical approach mentioned in Appendix A  of the [paper](http://cmph.sourceforge.net/papers/esa09.pdf) which uses a simpler hash but computes key hash only once as compared to twice for the random hash function approach. 

Changes from rkyv impl

1.  Different hash function 
2. Normally we would have stored the value inline at the slot, but here we store index to original placement of (k, v). 
3. `assignments.contains(&index)` in rkyv takes `O(bucket_chain_len)` which might slow hash building a bit (I haven't benchmarked it) depending on the bucket_chain len. So we use `generation` trick to get contains fast.

sffc commented:
Sounds good. When optimizing, please focus on (1) data size and (2) lookup speed. I don't care too much about building speed since it is done offline.

robertbastian commented on utils/zerovec/src/map/hashmap.rs:
```diff
@@ -0,0 +1,270 @@
+fn compute_hash<K: Hash>(seed: u32, k: K, m: usize) -> u32 {
+    let mut hasher = create_hasher_with_seed(seed.into());
+    k.hash(&mut hasher);
+    (hasher.finish() % m as u64) as u32
```
```suggestion
    (hasher.finish() as usize % m) as u32
```

This avoids doing a `u64` mod instruction on 32-bit architectures.

pdogr replied:
Done.

robertbastian replied:
I have to take this back, this does not give the same results on 32 and 64 bit. Assume the hash is 2^32 + 1, then on 32 bit that gets truncated to 1, which is not congruent to 2^32+1 under arbitrary moduli.

I don't think we need 64 bits here, so let's do the arithmetic in 32 bits and then widen to usize.

robertbastian commented on utils/zerovec/src/map/hashmap.rs:
```diff
@@ -0,0 +1,270 @@
+        let iter_len = iter.len();
+
+        // A vector to track the size of buckets for sorting.
+        let mut bucket_sizes = from_elem(0, iter_len);
```
For 'static' functions the style recommendation is to qualify them with one level of context (`vec::from_elem`).

That said, `from_elem` seems to be `#[doc(hidden)]` and I cannot figure out what the arguments are supposed to do. Is this just `vec![0; iter_len]`?

pdogr replied:
Yes I think this is what `vec!` macro uses internally. https://doc.rust-lang.org/src/alloc/macros.rs.html#42-55

robertbastian replied:
Please use the macro then, that's what everyone knows and understands.

robertbastian replied:
I recommend importing the macro with `use alloc::vec;`. This is what we do in other places as well, and the `#[macro_use]` attribute is like a wildcard import, which is something we want to avoid.

pdogr replied:
Done.

robertbastian commented on utils/zerovec/src/map/hashmap.rs:
```diff
@@ -0,0 +1,270 @@
+    }
+
+    #[inline]
+    pub fn build_from_exact_iter_with_hash_fn<K, I, A, H>(iter: I, h: H) -> Self
```
Nit: give `iter` a better name, like `keys`.

pdogr replied:
Done.

robertbastian commented on utils/zerovec/src/map/hashmap.rs:
```diff
@@ -0,0 +1,270 @@
+}
+
+#[inline]
+fn compute_hash<K: Hash>(seed: u32, k: K, m: usize) -> u32 {
```
In `Index` the result of `compute_hash` is always cast to `usize`. Why is it returning `u32`?

pdogr replied:
Yes realized this, changed the return type to usize.

robertbastian reviewed (commented):
I haven't reviewed the construction algorithm but wanted to leave my comments so far

pdogr commented on utils/zerovec/src/map/hashmap.rs:
```diff
@@ -16,40 +16,40 @@ fn create_hasher_with_seed(seed: u128) -> AHasher {
     let mut hasher = create_hasher_with_seed(seed.into());
     k.hash(&mut hasher);
-    (hasher.finish() % m as u64) as u32
+    (hasher.finish() as usize % m) as usize
```
Done.

pdogr commented:
Added some benchmarks reusing the data from the existing ones. 
Results
```
zeromap/lookup/large    time:   [188.63 ns 188.94 ns 189.27 ns]
                        change: [-2.8188% -2.3026% -1.7709%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 10 outliers among 100 measurements (10.00%)
  1 (1.00%) low mild
  3 (3.00%) high mild
  6 (6.00%) high severe

zerohashmap/lookup/large
                        time:   [82.513 ns 82.638 ns 82.766 ns]
                        change: [-1.1974% -0.7730% -0.3681%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 6 outliers among 100 measurements (6.00%)
  2 (2.00%) low mild
  1 (1.00%) high mild
  3 (3.00%) high severe

zeromap/lookup/large/hashmap
                        time:   [65.359 ns 65.476 ns 65.601 ns]
                        change: [-1.3054% -0.8122% -0.3220%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 8 outliers among 100 measurements (8.00%)
  2 (2.00%) low severe
  1 (1.00%) low mild
  4 (4.00%) high mild
  1 (1.00%) high severe
```
A few tradeoffs and observations  
1. Almost half of the time is used in the hash computation, I'll try "Practical approach" given in the above paper to see if it shows improvement. 
2. FlexZeroVec vs ZeroVec for HashIndex entries (size vs lookup time)
3. Reverse mapping. Ideally we want to store key, value together as `(K, V)` which would remove the need for reverse mapping and also better locality. 
4. Hash function. These benches are using `ahash` for now.  I will do some experiments with platform independent hashes.

sffc commented:
@pdogr Good data on the large map. How does the performance compare on the small version of the map?

pdogr commented:
> @pdogr Good data on the large map. How does the performance compare on the small version of the map?

For smaller keys hash computation time really comes into picture 

```
zeromap/lookup/small    time:   [52.080 ns 52.323 ns 52.614 ns]
                        change: [+5.7258% +6.4265% +7.0869%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 10 outliers among 100 measurements (10.00%)
  2 (2.00%) low mild
  4 (4.00%) high mild
  4 (4.00%) high severe

zerohashmap/lookup/small
                        time:   [98.832 ns 99.025 ns 99.231 ns]
                        change: [-0.8245% -0.3353% +0.1726%] (p = 0.19 > 0.05)
                        No change in performance detected.
Found 8 outliers among 100 measurements (8.00%)
  1 (1.00%) low mild
  5 (5.00%) high mild
  2 (2.00%) high severe

zeromap/lookup/small/hashmap
                        time:   [65.108 ns 65.336 ns 65.587 ns]
                        change: [-5.1179% -4.3311% -3.5866%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  2 (2.00%) low mild
  3 (3.00%) high mild
  3 (3.00%) high severe
```

pdogr commented:
FlexZeroVec vs ZeroVec
 
Using `ZeroVec` improves the lookup time by ~ 11% (99ns -> 87ns) for small lookups and ~25% (82ns -> 64ns) for large lookups with the existing algorithm.

sffc commented:
It seems that another way to architect this could be to make this a standalone type that maps from keys to an index, and then keep a proper ZeroMap as the second stage of lookup.

robertbastian commented on utils/zerovec/src/map/hashmap.rs:
```diff
@@ -0,0 +1,280 @@
+    where
+        A: Borrow<K>,
+        B: Borrow<V>,
+        I: ExactSizeIterator<Item = (A, B)>,
```
You shouldn't have to require an `ExactSizeIterator`. You can use `.size_hint()` to preallocate the containers, if the size hint is exact that's nice, if it's not then it's not the end of the world either. You can then make this implement `FromIterator`.

pdogr replied:
No this isn't for preallocating the container. We need the exact size for hash computation `hasher.finish() as usize % m`.

robertbastian replied:
But you have that after collecting into `K::Container` even if the iterator wasn't exact.

pdogr replied:
Changed to `Iterator`.

pdogr commented on utils/zerovec/src/map/hashmap.rs:
```diff
@@ -0,0 +1,280 @@
+            // Compute first level hash of the key bytes.
+            // First level uses a seed value of 0.
+            let l1 = h(0x00, k.borrow(), len);
+            if let Some(v) = bucket_sizes.get_mut(l1 as usize) {
+                *v += 1;
+            }
```
This lead to a clippy error for me so I didn't use this and went with the `if let Some` approach. 
```
error[E0658]: attributes on expressions are experimental
  --> utils/zerovec/src/map/hashmap.rs:50:13
   |
50 |             #[allow(clippy::indexing_slicing)] // 0 <= l1 < len
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #15701 <https://github.com/rust-lang/rust/issues/15701> for more information
```

robertbastian replied:
Uh, annoying. We could put the allow on the method itself and just comment each indexing operation, it's quite clear where they are.

pdogr replied:
Added this to method itself with comments at each slicing, unwrap.

pdogr commented on utils/zerovec/src/map/hashmap.rs:
```diff
@@ -0,0 +1,280 @@
+                #[allow(clippy::unwrap_used)] // i is in 0..self.keys.len()
+                K::Container::zvl_get_as_t(self.keys.zvl_get(i).unwrap(), |k| key.eq(k)).then(|| i)
+            })
+            .and_then(|i| self.values.zvl_get(i))
```
Done.

pdogr commented on utils/zerovec/src/map/hashmap.rs:
```diff
@@ -0,0 +1,280 @@
+use core::hash::{Hash, Hasher};
+
+#[inline]
+fn create_hasher_with_seed(seed: u128) -> AHasher {
```
Done.

pdogr commented on utils/zerovec/src/map/hashmap.rs:
```diff
@@ -0,0 +1,280 @@
+}
+
+#[derive(Debug)]
+pub struct HashIndex<'a> {
```
Made private.

pdogr commented on utils/zerovec/src/map/hashmap.rs:
```diff
@@ -0,0 +1,280 @@
+        K: 'b + ?Sized + Hash,
+        I: ExactSizeIterator<Item = A>,
+    {
+        HashIndex::build_from_exact_iter_with_hash_fn(keys, |seed, k, len| {
```
Removing `build_from_exact_iter_with_hash_fn`. We will only have `build_from_exact_iter`.

pdogr commented on utils/zerovec/src/map/hashmap.rs:
```diff
@@ -0,0 +1,280 @@
+    }
+
+    #[inline]
+    pub fn build_from_exact_iter_with_hash_fn<'b, K, I, A, H>(keys: I, h: H) -> Self
```
Yes this was from before when we hadn't decided to use a fix hashing algorithm. This will be removed now.

pdogr commented on utils/zerovec/src/map/hashmap.rs:
```diff
@@ -0,0 +1,280 @@
+        while start < len {
+            // Bucket span with the same first level hash
+            #[allow(clippy::indexing_slicing)] // start is always within bounds of `bucket_flatten`
+            let l1 = bucket_flatten[start].0 as usize;
```
Done.

pdogr commented on utils/zerovec/src/map/hashmap.rs:
```diff
@@ -0,0 +1,280 @@
+                generation += 1;
+
+                for (_, (k, _)) in buckets {
+                    let displacement_idx = h(seed, k.borrow(), len) as usize;
```
Removed.

pdogr commented on utils/zerovec/src/map/hashmap.rs:
```diff
@@ -0,0 +1,280 @@
+            let l1 = bucket_flatten[start].0 as usize;
+            #[allow(clippy::indexing_slicing)] // l1 is always within bounds of `bucket_sizes`
+            let end = start + bucket_sizes[l1];
+            #[allow(clippy::indexing_slicing)] // start, end are always within bounds of
```
As `bucket_sizes.sum() == len`. I think a better comment would be that `end -1` is always in bounds so slicing is safe.

robertbastian commented on utils/zerovec/src/map/hashmap.rs:
```diff
@@ -0,0 +1,263 @@
+    where
+        K: Hash,
+    {
+        let l1 = compute_hash(0, k, self.displacements.len());
```
This will panic if the map is empty, as `compute_hash` will do `% 0`. Add that as a precondition to `compute_hash` and guard against it here.

robertbastian commented on utils/zerovec/src/map/hashmap.rs:
```diff
@@ -0,0 +1,263 @@
+            None
+        } else {
+            let hash = compute_hash(seed as u32, k, self.displacements.len());
+            self.reverse_mapping.get(hash).map(|i| i as usize)
```
IIUC `reverse_mapping` has the same length as `displacements`, `keys`, and `values`, so this `get` will always succeed.

Also, IIUC, `reverse_mapping` is a permutation, which you could apply to `values` during construction.

pdogr replied:
Applying this permutation to keys and values in HashIndex

pdogr commented:
Switched to the practical approach mentioned in that paper with 64 bit (16 bits for g, 24 bits for f0, f1).
Changed the hash function to [wyhash](https://github.com/wangyi-fudan/wyhash). 
<p> 
<details><summary> Benches </summary> <br/>


```
zeromap/lookup/small    time:   [50.796 ns 50.913 ns 51.042 ns]
                        change: [+0.1679% +0.7275% +1.3576%] (p = 0.02 < 0.05)
                        Change within noise threshold.
Found 11 outliers among 100 measurements (11.00%)
  1 (1.00%) low severe
  1 (1.00%) low mild
  2 (2.00%) high mild
  7 (7.00%) high severe

zeromap/lookup/large    time:   [196.49 ns 197.21 ns 198.13 ns]
                        change: [-7.7555% -6.8637% -5.9877%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 10 outliers among 100 measurements (10.00%)
  2 (2.00%) low severe
  1 (1.00%) low mild
  5 (5.00%) high mild
  2 (2.00%) high severe

zeromap/lookup/small/hashmap
                        time:   [69.802 ns 70.089 ns 70.439 ns]
                        change: [+2.9376% +3.6548% +4.3437%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 10 outliers among 100 measurements (10.00%)
  2 (2.00%) low severe
  8 (8.00%) high mild

zeromap/lookup/large/hashmap
                        time:   [66.232 ns 66.464 ns 66.749 ns]
                        change: [+0.4782% +1.0997% +1.7342%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 11 outliers among 100 measurements (11.00%)
  2 (2.00%) low severe
  3 (3.00%) high mild
  6 (6.00%) high severe

zerohashmap/lookup/small
                        time:   [45.760 ns 45.844 ns 45.928 ns]
                        change: [-1.8354% -1.3624% -0.9334%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 6 outliers among 100 measurements (6.00%)
  2 (2.00%) low severe
  4 (4.00%) high mild

zerohashmap/lookup/large
                        time:   [43.285 ns 43.356 ns 43.434 ns]
                        change: [+3.1555% +3.8198% +4.4009%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 7 outliers among 100 measurements (7.00%)
  1 (1.00%) low severe
  2 (2.00%) low mild
  2 (2.00%) high mild
  2 (2.00%) high severe


```
</details>
</p>

robertbastian commented on utils/zerovec/src/map/hashmap.rs:
```diff
@@ -0,0 +1,287 @@
+}
+
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct ZeroHashMapStatic<'a, K, V>
```
'static' usually refers to the `'static` lifetime in Rust, not immutability. In this case I'd just call this struct `ZeroHashMap`, although later we need an enum over `ZeroHashMapOwned` (this) and `ZeroHashMapBorrowed`.

pdogr replied:
Renamed to ZeroHashMap

robertbastian commented on utils/zerovec/src/map/hashmap.rs:
```diff
@@ -0,0 +1,287 @@
+    V: ZeroMapKV<'a> + ?Sized,
+{
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    index: HashIndex<'a>,
```
I don't see the advantage of splitting `HashIndex` into a different type. This makes both construction and indexing spread out over two types/functions, which is harder to follow.

pdogr commented on utils/zerovec/src/map/hashmap.rs:
```diff
@@ -0,0 +1,263 @@
+// This file is part of ICU4X. For terms of use, please see the file
```
Done.

Manishearth commented on utils/zerovec/src/hashmap/mod.rs:
```diff
@@ -0,0 +1,7 @@
+
+#[allow(clippy::module_inception)] // module is purely internal
+mod hashmap;
+pub use hashmap::ZeroHashMapStatic;
```
we should have the code that is in hashmap.rs live in mod.rs

(we can split it again later if we really need to)

Manishearth commented on utils/zerovec/src/lib.rs:
```diff
@@ -231,9 +233,9 @@ mod yoke_impls;
-pub use crate::map::map::ZeroMap;
 #[cfg(feature = "hashmap")]
-pub use crate::map::ZeroHashMapStatic;
+pub use crate::hashmap::ZeroHashMapStatic;
```
thought: in the long run we need to figure out where everything lives. I would like to eventually have a separate zeromap crate with all of the different map types, though ZeroMapKV and the original ZeroMap might still live here.

I think there's a good chance to build map types that don't need the KV magic and can work based off of _containers_ rather than using the gettable type, which is definitely nice and magical but unfortunately causes a lot of complexity.

pdogr commented on utils/zerovec/src/map/hashmap.rs:
```diff
@@ -0,0 +1,263 @@
+}
+
+#[derive(Debug)]
+struct HashIndex<'a> {
```
Done.

robertbastian commented on utils/zerovec/src/hashmap/mod.rs:
```diff
@@ -207,24 +61,19 @@ where
+    /// Given a `key` return the probable index of the key or [`None`] if the key is guaranteed to be absent.
     /// Another check to determine if the key matches the one at the probable index is required.
     #[inline]
     pub fn index<A>(&self, key: &A) -> Option<usize>
```
This feels like an implementation detail, not much a client can do this this without access to `keys` and `values`.

Try to match `ZeroMap`'s API by implementing methods like `contains_key`, `iter_keys`, `iter_values`, etc. Those should be fairly simple.

robertbastian commented on utils/zerovec/src/hashmap/mod.rs:
```diff
@@ -207,24 +61,19 @@ where
     {
-        let (g, f0, f1) = compute_hash(key, self.displacements.len());
+        let hash = compute_hash(key.borrow());
+        let (g, f0, f1) = split_hash64(hash, self.len() as u32);
```
`m` and `g` should always be `usize`, because they're a length and an index. Casting them around is confusing to the compiler (that might miss optimisations) and the reader.

pdogr replied:
Changed `m` to always be `usize`.

robertbastian commented on utils/zerovec/benches/zeromap.rs:
```diff
@@ -304,7 +304,7 @@ fn build_zerohashmap(large: bool) -> ZeroHashMap<'static, Index32Str, Index32Str
     }
 
-    ZeroHashMap::build_from_iter(kv.iter().map(|kv| (indexify(&kv.0), kv.1)))
+    ZeroHashMap::from_iter(kv.iter().map(|kv| (indexify(&kv.0), kv.1)))
```
```suggestion
    kv.iter().map(|kv| (indexify(&kv.0), kv.1)).collect()
```

robertbastian commented on utils/zerovec/src/hashmap/mod.rs:
```diff
@@ -18,7 +18,7 @@ pub use algorithms::*;
 ///
 /// let kv = vec![(0, "a"), (1, "b"), (2, "c")];
-/// let hashmap: ZeroHashMap<i32, str> = ZeroHashMap::build_from_iter(kv.into_iter());
+/// let hashmap: ZeroHashMap<i32, str> = ZeroHashMap::from_iter(kv.into_iter());
```
It's more idiomatic to use `collect` to access `FromIterator`:

```rust
let hashmap = kv.into_iter().collect::<ZeroHashMap<i32, str>>();
```

If you want this to look like a constructor, you could still shorten this:
```rust
let hashmap = ZeroHashMap::<i32, str>::from_iter(kv.into_iter());
```

Also multiple cases below.

robertbastian commented on utils/zerovec/src/hashmap/algorithms.rs:
```diff
@@ -0,0 +1,167 @@
+    if d == (0, 0) || m == 0 {
+        None
+    } else {
+        Some((f.1.wrapping_mul(d.0).wrapping_add(f.0).wrapping_add(d.1) % m) as usize)
```
Reiterating my previous comment because the code moved: `wrapping_` does (mod 2^32) arithmetic, but you want (mod m).

pdogr replied:
Done

pdogr replied:
Reverted back to `wrapping_mul` and `wrapping_add`.

robertbastian commented on utils/zerovec/src/hashmap/mod.rs:
```diff
@@ -0,0 +1,181 @@
+    /// placeholder.
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    displacements: ZeroVec<'a, (u32, u32)>,
+    keys: K::Container,
```
for the next pr: this doesn't borrow keys and values. You'll need a `ZeroHashMapBorrowed` and all the boilerplate like in `ZeroMap`...

robertbastian commented on utils/zerovec/src/hashmap/mod.rs:
````diff
@@ -0,0 +1,181 @@
+/// assert_eq!(hashmap.get(&2), Some("c"));
+/// assert_eq!(hashmap.get(&4), None);
+/// ```
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
````
for the next pr: `ZeroMap` has a special implementation to have readable JSON, you can probably steal that by just pointing it at `keys` and `values`.

robertbastian commented on utils/zerovec/src/lib.rs:
```diff
@@ -215,6 +215,8 @@ extern crate alloc;
 mod error;
 mod flexzerovec;
+#[cfg(feature = "hashmap")]
+mod hashmap;
```
pub

robertbastian commented on utils/zerovec/src/hashmap/mod.rs:
```diff
@@ -0,0 +1,181 @@
+use alloc::vec;
+use core::hash::Hash;
+
+mod algorithms;
+pub use algorithms::*;
```
```suggestion
pub mod algorithms;
use algorithms::*
```

pdogr commented on utils/zerovec/src/hashmap/mod.rs:
```diff
@@ -95,14 +102,65 @@ where
+        (0..self.len()).map(|index| {
+            (
+                #[allow(clippy::unwrap_used)] // index is in range
+                self.keys.zvl_get(index).unwrap(),
```
Yeah I think I'll implement `Deserialize` for `ZeroHashMap` to uphold this invariant.

pdogr replied:
Done.

pdogr commented on utils/zerovec/src/hashmap/mod.rs:
```diff
@@ -95,14 +102,65 @@ where
+where
+    K: ZeroMapKV<'a> + ?Sized,
+    V: ZeroMapKV<'a> + ?Sized,
+{
```
Will do this in a follow-up.

pdogr commented on utils/zerovec/src/hashmap/algorithms.rs:
```diff
@@ -39,12 +37,13 @@ pub fn compute_hash<K: Hash + ?Sized>(key: &K) -> u64 {
     } else {
-        Some((f.1 % m * d.0 as usize + f.0 + d.1 as usize) % m)
+        // f.0, f.1 are 24bits
+        let r = (f.1 as u64 * d.0 as u64) + (f.0 + d.1) as u64;
```
Changed this back to using `wrapping_mul` and `wrapping_add`

robertbastian commented on utils/zerovec/derive/src/make_varule.rs:
```diff
@@ -241,6 +254,8 @@ pub fn make_varule_impl(attr: AttributeArgs, mut input: DeriveInput) -> TokenStr
 
         #maybe_debug
+
+        #maybe_hash
```
I don't think this is resolved?

pdogr replied:
Yes resolving this.
